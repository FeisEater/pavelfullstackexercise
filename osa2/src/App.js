import React from 'react'

class App extends React.Component {
    constructor(props) {
      super(props)
      this.state = {
        persons: [
          { name: 'Arto Hellas', number: '0800123123', id: 0 }
        ],
        newName: '',
        newNumber: ''
      }
    }
  
    addEntry = (event) => {
        event.preventDefault()
        
        if (this.state.persons.find(person => person.name === this.state.newName) !== undefined) {
            alert('Already added ' + this.state.newName);
            return;
        }

        const personObject = {
            name: this.state.newName,
            number: this.state.newNumber,
            id: this.state.persons.length
        }
        
        const persons = this.state.persons.concat(personObject)
        
        this.setState({
            persons,
            newName: '',
            newNumber: ''
        })
    }

    handleNameChange = (event) => {
        this.setState({ newName: event.target.value })
    }

    handleNumberChange = (event) => {
        this.setState({ newNumber: event.target.value })
    }
      
    render() {
      return (
        <div>
          <h2>Puhelinluettelo</h2>
          <form onSubmit={this.addEntry}>
            <div>
              nimi: <input value={this.state.newName} onChange={this.handleNameChange} />
            </div>
            <div>
              numero: <input value={this.state.newNumber} onChange={this.handleNumberChange} />
            </div>
            <div>
              <button type="submit">lisää</button>
            </div>
          </form>
          <h2>Numerot</h2>
          <table><tbody>
              {this.state.persons.map(person => <Person key={person.id} person={person} />)}
          </tbody></table>
        </div>
      )
    }
  }

const Person = (props) => {
    return (
        <tr>
            <td>{props.person.name}</td>
            <td>{props.person.number}</td>
        </tr>
    )
}

export default App