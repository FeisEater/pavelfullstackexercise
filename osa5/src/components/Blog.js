import React from 'react'

class Blog extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      expanded: false
    }
  }

  toggleExpansion = () => {
    this.setState({expanded: !this.state.expanded})
  }

  render() {

    const blogStyle = {
      paddingTop: 10,
      paddingLeft: 2,
      border: 'solid',
      borderWidth: 1,
      marginBottom: 5
    }

    const blog = this.props.blog

    if (!this.state.expanded) {
        return (
          <div style={blogStyle} onClick={this.toggleExpansion}>
            {blog.title} {blog.author}
          </div>
        )
    }

    return (
      <div style={blogStyle}>
        <div onClick={this.toggleExpansion}>{blog.title} {blog.author}</div>
        <div>
          <a href={'//' + blog.url}>{blog.url}</a>
          <div>{blog.likes} likes <button onClick={() => this.props.handleLike(blog)}>Like</button></div>
          <div>Added by {blog.user == null ? '--' : blog.user.name}</div>
          <div><button onClick={() => this.props.handleDelete(blog)}>Delete</button></div>
        </div>
      </div>
    )
  }
}

export default Blog