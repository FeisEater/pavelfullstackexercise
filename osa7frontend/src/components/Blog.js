import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { likeBlog, deleteBlog } from '../reducers/blogReducer'

const DeleteButton = ({deleteFunc, show}) => {
  if (!show)
    return null
  return (
    <div>
      <button onClick={deleteFunc}>Delete</button>
    </div>
  )
}

DeleteButton.propTypes = {
  deleteFunc: PropTypes.func.isRequired,
  show: PropTypes.bool.isRequired
}

class Blog extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      expanded: false
    }
  }

  static propTypes = {
    blog: PropTypes.any.isRequired
  }

  toggleExpansion = () => {
    this.setState({expanded: !this.state.expanded})
  }

  deleteBlog = async (blog) => {
    if (!window.confirm("Poistetaanko " + blog.title + "?"))
      return
    await this.props.deleteBlog(this.props.blog)
  }

  render() {

    const blogStyle = {
      paddingTop: 10,
      paddingLeft: 2,
      border: 'solid',
      borderWidth: 1,
      marginBottom: 5
    }

    const blog = this.props.blog

    if (!this.state.expanded) {
        return (
          <div style={blogStyle} onClick={this.toggleExpansion} className="blogHeader">
            {blog.title} {blog.author}
          </div>
        )
    }

    return (
      <div style={blogStyle}>
        <div onClick={this.toggleExpansion} className="blogHeader">{blog.title} {blog.author}</div>
        <div className="details">
          <a href={'//' + blog.url} className="urlLink">{blog.url}</a>
          <div className="likes">{blog.likes} likes <button onClick={() => this.props.likeBlog(blog)}>Like</button></div>
          <div className="creator">Added by {blog.user == null ? '--' : blog.user.name}</div>
          <DeleteButton deleteFunc={() => this.deleteBlog(blog)} show={blog.user === null || blog.user.username === this.props.loggedIn.username} />
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    loggedIn: state.loggedIn
  }
}

const ConnectedBlog = connect(
  mapStateToProps,
  { likeBlog, deleteBlog }
)(Blog)

export default ConnectedBlog