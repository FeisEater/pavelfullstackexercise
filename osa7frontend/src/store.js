import { createStore, combineReducers, applyMiddleware } from 'redux'
import blogReducer from './reducers/blogReducer'
import notificationReducer from './reducers/notificationReducer'
import loginReducer from './reducers/loginReducer'
import thunk from 'redux-thunk'

const reducer = combineReducers({
  blogs: blogReducer,
  notification: notificationReducer,
  loggedIn: loginReducer
})

const store = createStore(reducer, applyMiddleware(thunk))

export default store