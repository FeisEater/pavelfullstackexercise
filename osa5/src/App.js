import React from 'react'
import Blog from './components/Blog'
import Notification from './components/Notification'
import Togglable from './components/Togglable'
import BlogForm from './components/BlogForm'
import blogService from './services/blogs'
import loginService from './services/login'

class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      blogs: [],
      username: '',
      password: '',
      user: null,
      newtitle: '',
      newauthor: '',
      newurl: '',
      message: null,
      messageIsError: false
    }
    this.msgTimeout = setTimeout(() => {}, 0)
  }

  componentDidMount() {
    blogService.getAll().then(blogs =>
      this.setState({ blogs })
    )
    const loggedUserJSON = window.localStorage.getItem('loggedUser')
    if (loggedUserJSON) {
      const user = JSON.parse(loggedUserJSON)
      this.setState({user})
      blogService.setToken(user.token)
    }  
  } 

  showMessage = (msg, isError) => {
    this.setState({
      message: msg,
      messageIsError: isError
    })
    clearTimeout(this.msgTimeout)
    this.msgTimeout = setTimeout(() => {
      this.setState({ message: null })
    }, 5000)
  }

  login = async (event) => {
    event.preventDefault()
    try{
      const user = await loginService.login({
        username: this.state.username,
        password: this.state.password
      })
      window.localStorage.setItem('loggedUser', JSON.stringify(user))
      this.setState({ username: '', password: '', user})
      blogService.setToken(user.token)
    } catch(exception) {
      this.showMessage('käyttäjätunnus tai salasana virheellinen', true)
    }
  }

  logout = event => {
    window.localStorage.removeItem('loggedUser')   
    this.setState({user: null})
    blogService.setToken(null)
    this.showMessage('kirjauduttu ulos', false)
  }

  handleLoginFieldChange = (event) => {
    this.setState({ [event.target.name]: event.target.value })
  }

  handleBlogFieldChange = (event) => {
    this.setState({ [event.target.name]: event.target.value })
  }
  
  createBlog = async (event) => {
    event.preventDefault()    
    try {
      const savedBlog = await blogService.createBlog({
        title: this.state.newtitle,
        author: this.state.newauthor,
        url: this.state.newurl,
        likes: 0
      })
      this.setState({
        newtitle: '',
        newauthor: '',
        newurl: '',
        blogs: this.state.blogs.concat(savedBlog)
      })
      this.showMessage('Blogi luotu', false)
      this.blogForm.toggleVisibility()
    } catch (exception) {
      console.log(exception)
      this.showMessage('Jokin meni päin persettä, lue logi', true)      
    }
  }

  render() {
    if (this.state.user === null) {
      return (
        <div>
          <Notification message={this.state.message} isError={this.state.messageIsError}/>
          <h2>Kirjaudu</h2>
  
          <form onSubmit={this.login}>
            <div>
              käyttäjätunnus
              <input
                type="text"
                name="username"
                value={this.state.username}
                onChange={this.handleLoginFieldChange}
              />
            </div>
            <div>
              salasana
              <input
                type="password"
                name="password"
                value={this.state.password}
                onChange={this.handleLoginFieldChange}
              />
            </div>
            <button type="submit">kirjaudu</button>
          </form>
        </div>
      )
    }
    
    return (
      <div>
        <Notification message={this.state.message} isError={this.state.messageIsError}/>
        <h2>blogs</h2>
        <div>{this.state.user.name} logged in <button onClick={this.logout}>Log out</button></div>
        <br/>
        <Togglable buttonLabel="uusi blogi" ref={component => this.blogForm = component}>
          <BlogForm
            title={this.state.newtitle}
            author={this.state.newauthor}
            url={this.state.newurl}
            handleChange={this.handleBlogFieldChange}
            handleSubmit={this.createBlog}
          />
        </Togglable>
        <br/>
        {this.state.blogs.map(blog => 
          <Blog key={blog._id} blog={blog}/>
        )}
      </div>
    );
  }
}

export default App;
