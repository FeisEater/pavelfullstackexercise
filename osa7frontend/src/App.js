import React from 'react'
import Blog from './components/Blog'
import Notification from './components/Notification'
import Togglable from './components/Togglable'
import BlogForm from './components/BlogForm'
import blogService from './services/blogs'
import loginService from './services/login'
import { connect } from 'react-redux'
import { blogInitialization } from './reducers/blogReducer'
import { loginLocalStorage, login, logout } from './reducers/loginReducer'

class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      username: '',
      password: ''
    }
  }

  componentDidMount = async () => {
    await this.props.loginLocalStorage()
    await this.props.blogInitialization()
  } 

  login = async (event) => {
    event.preventDefault()
    await this.props.login({
      username: this.state.username,
      password: this.state.password
    })
    if (this.props.loggedIn)
      this.setState({ username: '', password: '' })
  }

  handleLoginFieldChange = (event) => {
    this.setState({ [event.target.name]: event.target.value })
  }

  render() {
    if (this.props.loggedIn === null) {
      return (
        <div>
          <Notification />
          <h2>Kirjaudu</h2>
  
          <form onSubmit={this.login} className="loginForm">
            <div>
              käyttäjätunnus
              <input
                type="text"
                name="username"
                value={this.state.username}
                onChange={this.handleLoginFieldChange}
              />
            </div>
            <div>
              salasana
              <input
                type="password"
                name="password"
                value={this.state.password}
                onChange={this.handleLoginFieldChange}
              />
            </div>
            <button type="submit">kirjaudu</button>
          </form>
        </div>
      )
    }
    
    return (
      <div>
        <Notification />
        <h2>blogs</h2>
        <div className="userDiv">{this.props.loggedIn.name} logged in <button onClick={this.props.logout}>Log out</button></div>
        <br/>
        <Togglable buttonLabel="uusi blogi" ref={component => this.blogForm = component}>
          <BlogForm blogForm={this.blogForm} />
        </Togglable>
        <br/>
        {this.props.blogs.map(blog => 
          <Blog key={blog._id}
                blog={blog} />
        )}
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    blogs: state.blogs.sort((a, b) => b.likes - a.likes),
    loggedIn: state.loggedIn
  }
}

export default connect(
  mapStateToProps,
  { blogInitialization, loginLocalStorage, login, logout }
)(App)