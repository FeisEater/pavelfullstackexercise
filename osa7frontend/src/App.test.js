import React from 'react'
import { mount } from 'enzyme'
import App from './App'
import Blog from './components/Blog'
jest.mock('./services/blogs')
import blogService from './services/blogs'
import store from './store'
import { Provider } from 'react-redux'

describe.only('<App />', () => {
  let app

  describe('when user is not logged', () => {
    beforeEach(() => {
        window.localStorage.removeItem('loggedUser')
        app = mount(<Provider store={store}><App /></Provider>)
    })

    it('only login form is rendered', () => {
      app.update()
      const blogComponents = app.find(Blog)
      expect(app.find('.loginForm').length).toEqual(1)
      expect(blogComponents.length).toEqual(0)
    })
  })

  describe('when user is logged', () => {
    beforeEach(() => {
      window.localStorage.setItem('loggedUser', JSON.stringify({
        token: 'token',
        username: 'username',
        name: 'Real Name'
      }))
      app = mount(<Provider store={store}><App /></Provider>)
    })

    it('all blogs are rendered', () => {
      app.update()
      const blogComponents = app.find(Blog)
      expect(app.find('.loginForm').length).toEqual(0)
      expect(blogComponents.length).toEqual(2)
    })
  })
})
