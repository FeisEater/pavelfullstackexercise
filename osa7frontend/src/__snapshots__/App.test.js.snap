// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> when user is logged snapshot test 1`] = `
<div>
  <h2>
    blogs
  </h2>
  <div
    className="userDiv"
  >
    Real Name
     logged in 
    <button
      onClick={[Function]}
    >
      Log out
    </button>
  </div>
  <br />
  <div>
    <div
      style={
        Object {
          "display": "",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        uusi blogi
      </button>
    </div>
    <div
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div>
        <h2>
          Create new
        </h2>
        <form
          onSubmit={[Function]}
        >
          <div>
            otsikko
            <input
              name="newtitle"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div>
            tekijä
            <input
              name="newauthor"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div>
            url
            <input
              name="newurl"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <button
            type="submit"
          >
            luo uusi
          </button>
        </form>
      </div>
      <button
        onClick={[Function]}
      >
        cancel
      </button>
    </div>
  </div>
  <br />
  <div
    className="blogHeader"
    onClick={[Function]}
    style={
      Object {
        "border": "solid",
        "borderWidth": 1,
        "marginBottom": 5,
        "paddingLeft": 2,
        "paddingTop": 10,
      }
    }
  >
    authorised title
     
    me
  </div>
  <div
    className="blogHeader"
    onClick={[Function]}
    style={
      Object {
        "border": "solid",
        "borderWidth": 1,
        "marginBottom": 5,
        "paddingLeft": 2,
        "paddingTop": 10,
      }
    }
  >
    otsikko
     
    tekijä
  </div>
</div>
`;

exports[`<App /> when user is not logged snapshot test 1`] = `
<div>
  <h2>
    Kirjaudu
  </h2>
  <form
    className="loginForm"
    onSubmit={[Function]}
  >
    <div>
      käyttäjätunnus
      <input
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div>
      salasana
      <input
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      kirjaudu
    </button>
  </form>
</div>
`;
